config/configgates/flags.go

```diff
@@ -0,0 +1,107 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package configgates
+
+import (
+	"flag"
+	"fmt"
+	"sort"
+	"strings"
+)
+
+const gatesListCfg = "feature-gates"
+
+var gatesList = FlagValue{}
+
+// Flags adds CLI flags for managing feature gates to the provided FlagSet.
+func Flags(flags *flag.FlagSet) {
+	flags.Var(
+		gatesList,
+		gatesListCfg,
+		"Comma-delimited list of feature gate identifiers. Prefix with '-' to disable the feature.  '+' or no prefix will enable the feature.")
+}
+
+// GetFlags returns the FlagValue used with Flags().
+func GetFlags() FlagValue {
+	return gatesList
+}
+
+var _ flag.Value = (*FlagValue)(nil)
+
+// FlagValue implements the flag.Value interface and provides a mechanism for applying feature
+// gate statuses to a Registry.
+type FlagValue map[string]bool
+
+// String returns a string representing the FlagValue.
+func (f FlagValue) String() string {
+	var t []string
+	for k, v := range f {
+		if v {
+			t = append(t, k)
+		} else {
+			t = append(t, "-"+k)
+		}
+	}
+
+	// Sort the list of identifiers for consistent results
+	sort.Strings(t)
+	return strings.Join(t, ",")
+}
+
+// Set applies the FlagValue encoded in the input string.
+func (f FlagValue) Set(s string) error {
+	return f.SetSlice(strings.Split(s, ","))
+}
+
+// SetSlice applies the feature gate statuses in the input slice to the FlagValue.
+func (f FlagValue) SetSlice(s []string) error {
+	for _, v := range s {
+		var id string
+		var val bool
+		switch v[0] {
+		case '-':
+			id = v[1:]
+			val = false
+		case '+':
+			id = v[1:]
+			val = true
+		default:
+			id = v
+			val = true
+		}
+
+		if _, exists := f[id]; exists {
+			// If the status has already been set, ignore it
+			// This allows CLI flags, which are processed first
+			// to take precedence over config settings
+			continue
+		}
+		f[id] = val
+	}
+
+	return nil
+}
+
+// Apply sets the feature gate statuses encoded in FlagValue on the provided Registry.
+func (f FlagValue) Apply(r *Registry) error {
```

Comment:
Should this be merged with "Frozen" so user has to call only one func, pass the configuration to the registry and get the "frozen Gates" specific to that configuration?

Reply:
Part of https://github.com/open-telemetry/opentelemetry-collector/issues/3957 we try to decouple the features that we have from the method of configuring them: flags, env var, etc.

It would be very nice for the "gates" to only work with the gates settings (similar to ServiceTelemetryLogsSettings, etc), and the flag is just a way to get the "Settings".

Reply:
> Should this be merged with "Frozen" so user has to call only one func, pass the configuration to the registry and get the "frozen Gates" specific to that configuration?

I think this can work.  I'll give it a try.  It might also help with the concern about using a `Registry` for testing.  If the state of a `Registry` doesn't need to be mutated to get the final frozen set, but instead just holds default values for registered gates, then test configurations can be applied to retrieve as many different frozen sets as are required.

I do worry that it will prevent the use of the unexported `*Gate` pattern in the README and that I used with the service telemetry gate in the original PR.  In that case I think it can be solved by passing the service config to the `colTelemetry.init()` method, but I wonder if there are other use cases that would be made significantly more complicated by this change.